components/script/dom/bluetooth.rs

```diff
@@ -49,9 +52,77 @@ const NAME_PREFIX_ERROR: &'static str = "'namePrefix', if present, must be nonem
 const NAME_TOO_LONG_ERROR: &'static str = "A device name can't be longer than 248 bytes.";
 const SERVICE_DATA_ERROR: &'static str = "'serviceData', if present, must be non-empty to filter devices.";
 const SERVICE_ERROR: &'static str = "'services', if present, must contain at least one service.";
-const OPTIONS_ERROR: &'static str = "Fields of 'options' conflict with each other.
+pub const OPTIONS_ERROR: &'static str = "Fields of 'options' conflict with each other.
  Either 'acceptAllDevices' member must be true, or 'filters' member must be set to a value.";
 
+
+thread_local!(pub static EXTRA_PERMISSION_DATA: RefCell<BluetoothPermissionData> =
+              RefCell::new(BluetoothPermissionData { allowedDevices: Vec::new() }));
+
+/*fn add_new_allowed_device(allowed_device: AllowedBluetoothDevice) {
+    EXTRA_PERMISSION_DATA.with(|epd| {
+        epd.borrow_mut().allowedDevices.push(allowed_device);
+    });
+}
+
+fn remove_allowed_device(device_id: DOMString) {
+    EXTRA_PERMISSION_DATA.with(|epd| {
+        epd.borrow_mut().allowedDevices.retain(|d| d.deviceId != device_id);
+    });
+}
+
+fn find_allowed_device(device_id: DOMString) -> Option<AllowedBluetoothDevice> {
+    EXTRA_PERMISSION_DATA.with(|epd| {
+        epd.borrow().allowedDevices.iter().find(|d| d.deviceId == device_id).cloned()
+    })
+}*/
+
+pub fn get_allowed_devices() -> Vec<AllowedBluetoothDevice> {
+    EXTRA_PERMISSION_DATA.with(|epd| {
+        epd.borrow().allowedDevices.clone()
```

Comment:
I can't find a way to get a `&'a Vec<AllowedBluetoothDevice>` or a `&'static Vec<AllowedBluetoothDevice>` from `EXTRA_PERMISSION_DATA `, that is why there are so many `impl Clone for ...`. The `borrow` is needed to get the `BluetoothPermissionData` (which contains the `Vec<AllowedBluetoothDevice> `) out from the `RefCell`.
But I am open to any idea :) .